### PR TITLE
USEID-869: Implement WebAuthn registration and authentication

### PR DESCRIPTION
**Pull request by mpanne**
# Secure device switch by implementing WebAuthn (Prototype)

## Goal
We believe that WebAuthn would bring a secure and user friendly solution to transfer information between mobile app and desktop app. Therefore, we would like to implement WebAuthn and try out, how it works and behaves.

## Description
Add package `credentials` which includes a new controller which handles requests from widget and eID-Client to register new credentials and authenticate for existing credentials.

## Note
This is a technical prototype which is hidden behind a feature flag. The code is not production ready and no tests exist.

### CODE REVIEW

File: src/main/kotlin/de/bund/digitalservice/useid/credentials/CredentialController.kt [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+internal const val CREDENTIALS_BASE_PATH = "/api/v1/credentials"
+
+@RestController
+@ConditionalOnProperty(name = ["features.desktop-solution-enabled"], havingValue = "true")
+class CredentialController(
+    private val credentialService: CredentialService,
+    private val eventService: EventService,
+) {
+    @PostMapping(path = [CREDENTIALS_BASE_PATH])
```
Comment (code28) [inline src/main/kotlin/de/bund/digitalservice/useid/credentials/CredentialController.kt:25]: Let's add correct swagger docs right away:
```
     @ResponseStatus(HttpStatus.ACCEPTED)
     @Operation(summary = "Start registration for webauthn credentials")
     @ApiResponse(responseCode = "202", content = [Content()])
```

File: src/main/kotlin/de/bund/digitalservice/useid/credentials/CredentialController.kt [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+                    credential.credentialId,
+                    credential.pckCreationOptions.toCredentialsCreateJson(),
+                ),
+            )
+    }
+
+    @PutMapping(
+        path = ["$CREDENTIALS_BASE_PATH/{credentialId}"],
+        headers = ["Content-Type=application/json"],
+    )
```
Comment (code28) [inline src/main/kotlin/de/bund/digitalservice/useid/credentials/CredentialController.kt:48]: (no content of course only when the code is adjusted..)
```
     @ResponseStatus(HttpStatus.NO_CONTENT)
     @Operation(summary = "Complete webauthn registration")
     @ApiResponse(responseCode = "204", content = [Content()])
```

File: src/main/kotlin/de/bund/digitalservice/useid/credentials/CredentialController.kt [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+            startRegistrationRequest.refreshAddress,
+        )
+
+        return ResponseEntity
+            .status(HttpStatus.ACCEPTED)
+            .contentType(MediaType.APPLICATION_JSON)
+            .body(
+                StartRegistrationResponse(
+                    credential.credentialId,
+                    credential.pckCreationOptions.toCredentialsCreateJson(),
```
Comment (code28) [inline src/main/kotlin/de/bund/digitalservice/useid/credentials/CredentialController.kt:40]:
```suggestion
                    credential.pkcCreationOptions.toCredentialsCreateJson(),
```

File: src/main/kotlin/de/bund/digitalservice/useid/credentials/CredentialService.kt [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+        credentialId: UUID,
+        publicKeyCredentialJson: String,
+    ): Credential {
+        val credential = credentialMockDatasource.findById(credentialId)
+            ?: throw CredentialNotFoundException(credentialId)
+
+        val pkc = PublicKeyCredential.parseRegistrationResponseJson(publicKeyCredentialJson)
+
+        val finishRegistrationOptions = FinishRegistrationOptions.builder()
+            .request(credential.pckCreationOptions)
```
Comment (code28) [inline src/main/kotlin/de/bund/digitalservice/useid/credentials/CredentialService.kt:81]:
```suggestion
            .request(credential.pkcCreationOptions)
```

File: src/main/resources/static/js/webauthn.js [hunk 5]
```diff
@@ -0,0 +1,145 @@
+function registerWebAuthnCredentials(widgetSessionId) {
```
Comment (code28) [inline src/main/resources/static/js/webauthn.js:1]: Shall we put a divider commment between registration (app simulator) and authentication (widget)?

**Review by code28 — CHANGES_REQUESTED** [review]
Really nice! 🚀
And very clean code, I like it!

File: src/main/kotlin/de/bund/digitalservice/useid/credentials/CredentialController.kt [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+        @RequestBody publicKeyCredentialJson: String,
+    ): ResponseEntity<Any> {
+        credentialService.finishRegistration(credentialId, publicKeyCredentialJson)
+
+        val credential = credentialService.startAuthentication(credentialId)
+        sendAuthenticateEventToWidget(credential)
+
+        return ResponseEntity
+            .status(HttpStatus.ACCEPTED)
+            .contentType(MediaType.APPLICATION_JSON)
```
[the remark this answers is not in the document]
Comment (hannesschaletzky) [reply src/main/kotlin/de/bund/digitalservice/useid/credentials/CredentialController.kt:60]: +1

File: src/main/kotlin/de/bund/digitalservice/useid/credentials/CredentialController.kt [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+            .build()
+    }
+
+    private fun sendAuthenticateEventToWidget(credential: Credential) {
+        eventService.publish(
+            AuthenticateEvent(credential.credentialId, credential.assertionRequest!!.toCredentialsGetJson()),
```
Comment (hannesschaletzky) [inline src/main/kotlin/de/bund/digitalservice/useid/credentials/CredentialController.kt:66]: I don't have a suggestion, but maybe we could get rid of the double bang operator.

File: src/main/kotlin/de/bund/digitalservice/useid/credentials/CredentialService.kt [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+    private fun buildUser(widgetSessionId: UUID): UserIdentity {
+        val username = widgetSessionId.toString()
+        return UserIdentity.builder()
+            .name(username)
+            .displayName(username)
+            .id(generateUserId())
+            .build()
+    }
+
+    private fun buildStartRegistrationOptions(user: UserIdentity): StartRegistrationOptions? {
```
Comment (hannesschaletzky) [inline src/main/kotlin/de/bund/digitalservice/useid/credentials/CredentialService.kt:60]: Why does this have an optional return type?

File: src/main/kotlin/de/bund/digitalservice/useid/credentials/CredentialService.kt [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+
+        return credentialMockDatasource.update(credential)
+    }
+
+    fun startAuthentication(credentialId: UUID): Credential {
+        val credential = credentialMockDatasource.findById(credentialId)
+            ?: throw CredentialNotFoundException(credentialId)
+
+        val assertionRequest: AssertionRequest = relyingParty.startAssertion(
+            StartAssertionOptions.builder().username(credential.username).build(),
```
Comment (hannesschaletzky) [inline src/main/kotlin/de/bund/digitalservice/useid/credentials/CredentialService.kt:102]: can we also do that builder pattern with a new line each time a new function is called? "Inline" builder patterns look weird to me

File: src/main/resources/static/js/webauthn.js (hunk 5 above)
Comment (hannesschaletzky) [reply src/main/resources/static/js/webauthn.js:1]: +1

**Review by hannesschaletzky — APPROVED** [review]
Especially the CredentialController and the CredentialService reads like a charme 👍 very clean IMHO. 
I did not have a look onto the JS / app-simulator code.

File: src/main/resources/static/js/webauthn.js (hunk 5 above)
Comment (hannesschaletzky) [reply src/main/resources/static/js/webauthn.js:1]: +1

File: src/main/kotlin/de/bund/digitalservice/useid/credentials/Credential.kt [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+import com.yubico.webauthn.data.PublicKeyCredentialDescriptor
+import java.util.UUID
+
+data class Credential(
+    val credentialId: UUID,
+    val widgetSessionId: UUID,
+    val username: String,
+    val userIdBase64: String,
+    val refreshAddress: String,
+    val pckCreationOptions: PublicKeyCredentialCreationOptions,
```
[the remark this answers is not in the document]
Comment (mpanne) [reply src/main/kotlin/de/bund/digitalservice/useid/credentials/Credential.kt:15]: thx, done

File: src/main/kotlin/de/bund/digitalservice/useid/credentials/CredentialController.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (mpanne) [reply src/main/kotlin/de/bund/digitalservice/useid/credentials/CredentialController.kt:25]: Good catch, I cleaned that up.

[the remark this answers is not in the document]
Comment (mpanne) [reply src/main/kotlin/de/bund/digitalservice/useid/credentials/CredentialController.kt]: Meaning I added `produces` everywhere and removed `headers` which is set automatically returning a `ResponseEntity`.

File: src/main/kotlin/de/bund/digitalservice/useid/credentials/CredentialController.kt (hunk 6 above)
[the remark this answers is not in the document]
Comment (mpanne) [reply src/main/kotlin/de/bund/digitalservice/useid/credentials/CredentialController.kt:62]: Yes, we should return 204 (see [here](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/PUT)):

> If the target resource does have a current representation and that representation is successfully modified in accordance with the state of the enclosed representation, then the origin server must send either a [200](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/200) (OK) or a [204](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/204) (No Content) response to indicate successful completion of the request.

[the remark this answers is not in the document]
Comment (mpanne) [reply src/main/kotlin/de/bund/digitalservice/useid/credentials/CredentialController.kt]: Done

File: src/main/kotlin/de/bund/digitalservice/useid/credentials/CredentialMockDatasource.kt [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+        return credentials.find { it.keyId?.id == keyCredentialId }
+    }
+
+    fun delete(credentialId: UUID) {
+        credentials.removeAll { it.credentialId == credentialId }
+    }
+
+    override fun getCredentialIdsForUsername(username: String): MutableSet<PublicKeyCredentialDescriptor> {
+        val credential = findByUsername(username) ?: return mutableSetOf()
+        return mutableSetOf(credential.keyId!!)
```
[the remark this answers is not in the document]
Comment (mpanne) [reply src/main/kotlin/de/bund/digitalservice/useid/credentials/CredentialMockDatasource.kt:50]: What errors occurred when testing? Shouldn't the `keyId` always be present?

File: src/main/kotlin/de/bund/digitalservice/useid/credentials/CredentialController.kt (hunk 1 above)
Comment (mpanne) [reply src/main/kotlin/de/bund/digitalservice/useid/credentials/CredentialController.kt]: Thanks, will do. Returning 202 makes no sense here, right? The entity is created right away, so we should return 201.

Comment (mpanne) [reply src/main/kotlin/de/bund/digitalservice/useid/credentials/CredentialController.kt]: Done

File: src/main/kotlin/de/bund/digitalservice/useid/credentials/CredentialMockDatasource.kt (hunk 11 above)
[the remark this answers is not in the document]
Comment (code28) [reply src/main/kotlin/de/bund/digitalservice/useid/credentials/CredentialMockDatasource.kt:50]: I don't recall unfortunately. The `keyId` should be present actually - maybe it was a weird state, when testing. I'll see if I can reproduce it. But anyway, I think having as few force unwraps as possible is good, so I'd change that anyway. What do you think?

File: src/main/resources/static/js/webauthn.js (hunk 5 above)
Comment (mpanne) [reply src/main/resources/static/js/webauthn.js:1]: I'm totally not a friend of divider comments. Reasons: 
- You always forget them when refactoring / moving code, so they lose their meaning and are more confusing than helpful pretty soon.
- There are functions in this file which are used by both, registration and authentication. Another separator for that?

Comment (mpanne) [reply src/main/resources/static/js/webauthn.js:1]: If you wanna separate it, I could create multiple files. 
(Totally out of scope: Once we have a build pipeline for our widget, we would split up those files anyways and have them re-combined during the build)

File: src/main/kotlin/de/bund/digitalservice/useid/credentials/CredentialController.kt (hunk 7 above)
Comment (mpanne) [reply src/main/kotlin/de/bund/digitalservice/useid/credentials/CredentialController.kt:78]: Added a null check before, which throws a more descriptive exception.

File: src/main/kotlin/de/bund/digitalservice/useid/credentials/CredentialService.kt (hunk 8 above)
Comment (mpanne) [reply src/main/kotlin/de/bund/digitalservice/useid/credentials/CredentialService.kt:60]: Good catch. Auto extracted method. Removed.

File: src/main/kotlin/de/bund/digitalservice/useid/credentials/CredentialService.kt (hunk 9 above)
Comment (mpanne) [reply src/main/kotlin/de/bund/digitalservice/useid/credentials/CredentialService.kt:102]: Done

File: allowed-licenses.json [hunk 12]
```diff
@@ -89,6 +89,13 @@
     },
     {
       "moduleLicense": "BSD Zero Clause License"
+    },
+    {
+      "moduleLicense": "BSD3"
+    },
+    {
+      "moduleLicense": null,
+      "moduleName": "com.yubico:webauthn-server-parent"
```
[the remark this answers is not in the document]
Comment (mpanne) [reply allowed-licenses.json]: Declares to use BSD license (see [here](https://mvnrepository.com/artifact/com.yubico/webauthn-server-parent/2.4.0)). BSD is compatible to EUPL in all versions according to [this matrix](https://joinup.ec.europa.eu/collection/eupl/matrix-eupl-compatible-open-source-licences).

[the remark this answers is not in the document]
Comment (mpanne) [reply allowed-licenses.json]: I added a note

**Review by code28 — APPROVED** [review]
I'd still vote for removing force unwrap if possible, even if I can't reproduce the error, but for me it's okay to keep it like that, until we can reproduce it. 🙃

File: src/main/resources/static/js/webauthn.js (hunk 5 above)
Comment (mpanne) [reply src/main/resources/static/js/webauthn.js:1]: Got your point now. 😉 I will add comments to the methods only used for the app simulation.

**Pull request merged**